pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,226 @@
+			}
+			clusterWObj, err := lister.Get(cWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := clusterWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: cWObjRef.Group, Version: cWObjRef.Version, Kind: clusterWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   cWObjRef.Resource,
+					ObjectName: cache.NewObjectName("", cWObjRef.Name),
+				}
+				logger.V(2).Info("Found cluster scoped singleton workload object", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should have status synced", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
+					c.singletonState.set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object status syncing is not driven by Binding", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName, "binding", bdg.Name)
+					c.singletonState.addIfNotExist(wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	if !reconcile {
+		return nil
+	}
+
+	for wObjIdentifier := range c.singletonState.wObjSync {
+		if err := c.reconcileSingletonByWObj(ctx, wObjIdentifier); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) reconcileSingletonByWObj(ctx context.Context, wObjID util.ObjectIdentifier) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Reconciling singleton workload object", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+
+	if !c.singletonState.get(wObjID) {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	list, _ := c.workStatusLister.ByNamespace("").List(labels.Everything())
+	for _, obj := range list {
+		sourceRef, err := util.GetSourceRefFromWorkStatus(obj)
+		if err != nil {
+			return err
+		}
+		sourceObjID := util.ObjectIdentifierFromSourceRef(sourceRef)
+		if !reflect.DeepEqual(wObjID, sourceObjID) {
+			continue
+		}
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		wsRef, _ := runtimeObjectToWorkStatusRef(obj)
+		logger.V(2).Info("Updating singleton status", "workstatus", wsRef.Name, "wecName", wsRef.WECName)
+		return updateObjectStatus(ctx, &wsRef.SourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+	}
+	return nil
+}
+
+func (c *Controller) reconcileSingletonByBdg(ctx context.Context, ref bindingRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Reconciling singleton status by binding", "name", string(ref))
+	return c.buildSingletonStateAndOptionallyReconcile(ctx, true) // TODO: differentiate full build and incremental change
```

Comment:
Yeah. It is not only a gross matter of efficiency, it is also a matter of correctness: this line can be run concurrently for multiple different values of `ref bindingRef` (which it discards because it does a full build). And further, each full build finishes by reconciling for a bunch of workload objects.